src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -0,0 +1,345 @@
+        {
+            runningInVisualStudio = false;
+
+            // Check to see if we're running inside of Visual Studio
+            foreach (var process in processNames)
+            {
+                if (IsProcessInList(process, s_visualStudioProcess))
+                {
+                    runningInVisualStudio = true;
+
+                    // This assumes running from VS\Common7\IDE\<process>.exe.
+                    return FileUtilities.GetFolderAbove(process, 3);
+                }
+            }
+
+            // VSInstallDir is set from the Developer Command Prompt
+            var vsInstallDir = Environment.GetEnvironmentVariable("VSINSTALLDIR");
+            var vsVersion = Environment.GetEnvironmentVariable("VisualStudioVersion");
+
+            if (!string.IsNullOrEmpty(vsInstallDir) &&
+                !string.IsNullOrEmpty(vsVersion) &&
+                vsVersion == "15.0" &&
+                Directory.Exists(vsInstallDir))
+            {
+                return vsInstallDir;
+            }
+
+            // Check assuming we're running in VS\MSBuild\15.0\Bin\MSBuild.exe
+            foreach (var process in processNames)
+            {
+                if (IsProcessInList(process, s_msBuildProcess))
+                {
+                    var vsPath = FileUtilities.GetFolderAbove(process, 4);
+                    var devEnv = FileUtilities.CombinePaths(vsPath, "Common7", "IDE", "devenv.exe");
+
+                    // Make sure VS is actually there before we suggest this root.
+                    if (File.Exists(devEnv))
+                    {
+                        return vsPath;
+                    }
+
+                    // VS\MSBuild\15.0\Bin\amd64\MSBuild.exe
+                    var vsPath64 = FileUtilities.GetFolderAbove(process, 5);
+                    var devEnv64 = FileUtilities.CombinePaths(vsPath64, "Common7", "IDE", "devenv.exe");
+
+                    // Make sure VS is actually there before we suggest this root.
+                    if (File.Exists(devEnv64))
+                    {
+                        return vsPath64;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns true if processName appears in the processList
+        /// </summary>
+        /// <param name="processName">Name of the process</param>
+        /// <param name="processList">List of processes to check</param>
+        /// <returns></returns>
+        private static bool IsProcessInList(string processName, string[] processList)
+        {
+            return processList.Any(s => Path.GetFileNameWithoutExtension(processName)?.IndexOf(s, StringComparison.OrdinalIgnoreCase) >= 0);
+        }
+
+        private static string GetProcessFromCommandLine()
+        {
+#if FEATURE_GET_COMMANDLINE
+            return Environment.GetCommandLineArgs()[0];
+#else
+            return null;
+#endif
```

Comment:
I checked where that null may flow into, and it appears to only flow into Path.GetFileNameWithoutExtension, which seems to return empty string for null.